File: third-party-service/src/model/proposal.ts
import { Author } from "./autor";
import { Product } from "./product";

export class Proposal {

    private readonly product: Product;
    private readonly bidValue: number;
    private readonly author: Author;

    constructor() {
        this.product = new Product();
        this.bidValue = this.generateBidValue();
        this.author = new Author();
    }

    getProduct(): Product {
        return this.product;
    }

    getBidValue(): number {
        return this.bidValue;
    }

    getAuthor(): Author {
        return this.author;
    }

    private generateBidValue(): number {
        const highValue: number = 25000;
        const lowValue: number = 15;
        const value: number = lowValue + Math.random() * (highValue - lowValue);
        return Number(value.toFixed(2));
    }
}

File: auction-consumer/src/main/java/com/lucasilva/auction_consumer/model/Proposal.java
package com.lucasilva.auction_consumer.model;

import jakarta.persistence.*;

import java.math.BigDecimal;

import static java.util.Objects.requireNonNull;

@Entity
public class Proposal {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(precision = 9, scale = 2)
    private BigDecimal value;

    private String author;

    private String email;

    @ManyToOne
    @JoinColumn(name = "product_id")
    private Product product;

    public Proposal() {}

    private Proposal(BigDecimal value, String author, String email) {
        this.value = value;
        this.author = author;
        this.email = email;
    }

    public static Proposal toProposal(avro.kafka.model.Proposal proposalEvent) {
        return new Proposal(
                proposalEvent.getBidValue(),
                proposalEvent.getAuthor().getName(),
                proposalEvent.getAuthor().getEmail());
    }

    public Long getId() {
        return id;
    }

    public BigDecimal getValue() {
        return value;
    }

    public String getAuthor() {
        return author;
    }

    public String getEmail() {
        return email;
    }

    public Product getProduct() {
        return product;
    }

    public void linkToProduct(Product product) {
        requireNonNull(product, "product is required!");
        this.product = product;
    }
}
